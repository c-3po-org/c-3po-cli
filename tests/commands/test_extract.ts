// import '../fixtures/declarations';
import * as path from "path";
import * as fs from "fs";
import { execSync } from "child_process";

const potPath = path.resolve(__dirname, "../../dist/translation.pot");
const baseTestPath = path.resolve(__dirname, "../fixtures/baseTest");
const sortByMsgidPath = path.resolve(__dirname, "../fixtures/sortByMsgidTest");
const ukTestPath = path.resolve(__dirname, "../fixtures/ukLocaleTest");
const jsxPath = path.resolve(__dirname, "../fixtures/testJSXParse.jsx");
const globalFn = path.resolve(__dirname, "../fixtures/globalFunc.js");
const tsPath = path.resolve(__dirname, "../fixtures/tSParse.ts");
const tsChaning = path.resolve(__dirname, "../fixtures/tsOptionalChaning.ts");
const tsxPath = path.resolve(__dirname, "../fixtures/tSXParse.tsx");

function cleanup() {
    fs.unlinkSync(potPath);
}

afterEach(() => {
    cleanup();
});

test("extract base case", () => {
    execSync(`ts-node src/index.ts extract -o ${potPath} ${baseTestPath}`);
    const result = fs.readFileSync(potPath).toString();
    expect(result).toMatchSnapshot();
});

test("extract from jsx", () => {
    execSync(`ts-node src/index.ts extract -o ${potPath} ${jsxPath}`);
    const result = fs.readFileSync(potPath).toString();
    expect(result).toMatchSnapshot();
});

test("extract from ts", () => {
    execSync(`ts-node src/index.ts extract -o ${potPath} ${tsPath}`);
    const result = fs.readFileSync(potPath).toString();
    expect(result).toMatchSnapshot();
});

test("extract from tsx", () => {
    execSync(`ts-node src/index.ts extract -o ${potPath} ${tsxPath}`);
    const result = fs.readFileSync(potPath).toString();
    expect(result).toMatchSnapshot();
});

test("extract from js with another default locale", () => {
    execSync(`ts-node src/index.ts extract -l uk -o ${potPath} ${ukTestPath}`);
    const result = fs.readFileSync(potPath).toString();
    expect(result).toMatchSnapshot();
});

test("should override babel plugin settings", () => {
    execSync(
        `ts-node src/index.ts extract --discover=_ -o ${potPath} ${globalFn}`
    );
    const result = fs.readFileSync(potPath).toString();
    expect(result).toContain('msgid "test _"');
});

test("should extract with numberedExpressions", () => {
    execSync(
        `ts-node src/index.ts extract --numberedExpressions -o ${potPath} ${baseTestPath}`
    );
    const result = fs.readFileSync(potPath).toString();
    expect(result).toContain("test translation 2 ${ 0 }");
});

test("should extract with extract.location", () => {
    execSync(
        `ts-node src/index.ts extract --extractLocation=never -o ${potPath} ${baseTestPath}`
    );
    const result = fs.readFileSync(potPath).toString();
    expect(result).not.toContain("#: tests/fixtures/baseTest");
});

test("should extract in the alphabetical order (sortByMsgid)", () => {
    execSync(
        `ts-node src/index.ts extract --sortByMsgid -o ${potPath} ${sortByMsgidPath}`
    );
    const result = fs.readFileSync(potPath).toString();
    expect(result).toMatchSnapshot();
});

test("extract from ts", () => {
    execSync(`ts-node src/index.ts extract -o ${potPath} ${tsChaning}`);
    const result = fs.readFileSync(potPath).toString();
    expect(result).toMatchSnapshot();
});
