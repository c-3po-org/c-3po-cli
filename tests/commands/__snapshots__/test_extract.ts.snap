// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`extract base case 1`] = `
"msgid \\"\\"
msgstr \\"\\"
\\"Content-Type: text/plain; charset=utf-8\\\\n\\"
\\"Plural-Forms: nplurals=2; plural=(n!=1);\\\\n\\"

#: tests/fixtures/baseTest/test1.js:4
#, javascript-format
msgid \\"test translation \${ name }\\"
msgstr \\"\\"

#: tests/fixtures/baseTest/nested/test2.js:4
msgid \\"test translation 2 \${ name }\\"
msgstr \\"\\""
`;

exports[`extract from js with another default locale 1`] = `
"msgid \\"\\"
msgstr \\"\\"
\\"Content-Type: text/plain; charset=utf-8\\\\n\\"
\\"Plural-Forms: nplurals=2; plural=(n!=1);\\\\n\\"

#: tests/fixtures/ukLocaleTest/index.js:5
#, javascript-format
msgid \\"\${ n } банан\\"
msgid_plural \\"\${ n } банана\\"
msgstr[0] \\"\\"
msgstr[1] \\"\\"
msgstr[2] \\"\\""
`;

exports[`extract from jsx 1`] = `
"msgid \\"\\"
msgstr \\"\\"
\\"Content-Type: text/plain; charset=utf-8\\\\n\\"
\\"Plural-Forms: nplurals=2; plural=(n!=1);\\\\n\\"

#: tests/fixtures/testJSXParse.jsx:6
msgid \\"test from jsx\\"
msgstr \\"\\""
`;

exports[`extract from ts 1`] = `
"msgid \\"\\"
msgstr \\"\\"
\\"Content-Type: text/plain; charset=utf-8\\\\n\\"
\\"Plural-Forms: nplurals=2; plural=(n!=1);\\\\n\\"

#: tests/fixtures/tSParse.ts:4
#, javascript-format
msgid \\"\${ a } string\\"
msgstr \\"\\""
`;

exports[`extract from ts 2`] = `
"msgid \\"\\"
msgstr \\"\\"
\\"Content-Type: text/plain; charset=utf-8\\\\n\\"
\\"Plural-Forms: nplurals=2; plural=(n!=1);\\\\n\\"

#: tests/fixtures/tsOptionalChaning.ts:5
#, javascript-format
msgid \\"test \${ name }\\"
msgstr \\"\\""
`;

exports[`extract from ts 3`] = `
"msgid \\"\\"
msgstr \\"\\"
\\"Content-Type: text/plain; charset=utf-8\\\\n\\"
\\"Plural-Forms: nplurals=2; plural=(n!=1);\\\\n\\"

#: tests/fixtures/tsNullishCoalescing.ts:5
#, javascript-format
msgid \\"test \${ name }\\"
msgstr \\"\\""
`;

exports[`extract from tsx 1`] = `
"msgid \\"\\"
msgstr \\"\\"
\\"Content-Type: text/plain; charset=utf-8\\\\n\\"
\\"Plural-Forms: nplurals=2; plural=(n!=1);\\\\n\\"

#: tests/fixtures/tSXParse.tsx:4
msgid \\"test from tsx\\"
msgstr \\"\\""
`;

exports[`extract from vue 1`] = `
"msgid \\"\\"
msgstr \\"\\"
\\"Content-Type: text/plain; charset=utf-8\\\\n\\"
\\"Plural-Forms: nplurals=2; plural=(n!=1);\\\\n\\"

#: tests/fixtures/testVueParse.vue:7
msgid \\"test from vue file\\"
msgstr \\"\\""
`;

exports[`should extract in the alphabetical order (sortByMsgid) 1`] = `
"msgid \\"\\"
msgstr \\"\\"
\\"Content-Type: text/plain; charset=utf-8\\\\n\\"
\\"Plural-Forms: nplurals=2; plural=(n!=1);\\\\n\\"

#: tests/fixtures/sortByMsgidTest/test.js:10
msgid \\"a translation \${ name }\\"
msgstr \\"\\"

#: tests/fixtures/sortByMsgidTest/test.js:4
#, javascript-format
msgid \\"c translation \${ name }\\"
msgstr \\"\\"

#: tests/fixtures/sortByMsgidTest/test.js:6
msgid \\"f translation \${ name }\\"
msgstr \\"\\"

#: tests/fixtures/sortByMsgidTest/test.js:8
#, javascript-format
msgid \\"z translation \${ name }\\"
msgstr \\"\\""
`;
